### PR TITLE
[FLINK-36379] Improve (Global)Committer with UC disabled

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

[FLINK-36287](https://issues.apache.org/jira/browse/FLINK-36287) disabled UC for all inter-sink connections to adhere to the contract of notifyCheckpointCompleted. This allows us to remove some special casing and improve global committer.

## Brief change log

* Refactor sink test assertions 
* Optimize global committers 
* Optimize committers with UC disabled

See commit messages for more details.

## Verifying this change

Covered by tests in
* org.apache.flink.streaming.runtime.operators.sink
* org.apache.flink.streaming.api.connector.sink2
* Various IT cases and E2E tests

I modified / extended the former two suites.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / no / **don't know**)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        transformation.setParallelism(1);
+        transformation.setMaxParallelism(1);
+        return Collections.emptyList();
+    }
+
+    /**
+     * Looks for a committer in the pipeline and aborts on writer. The GlobalCommitter behaves
+     * differently if there is a committer after the writer.
+     */
+    private static boolean hasUpstreamCommitter(DataStream<?> ds) {
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java:100]: How safe is this really? Users could have used the global committer with different custom operators in front of that follow either writer or committer semantics.

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManagerImpl.java [hunk 2]
```diff
@@ -168,13 +169,33 @@ public Collection<CommittableWithLineage<CommT>> commit(Committer<CommT> committ
...
+    Collection<CommitRequestImpl<CommT>> getPendingRequests(boolean assertFull) {
         return subtasksCommittableManagers.values().stream()
-                .filter(subtask -> !onlyIfFullyReceived || subtask.hasReceivedAll())
+                .peek(subtask -> assertReceivedAll(assertFull, subtask))
                 .flatMap(SubtaskCommittableManager::getPendingRequests)
                 .collect(Collectors.toList());
     }
 
+    /**
+     * Sinks don't use unaligned checkpoints, so we receive all committables of a given upstream
```
Comment (fapaul) [inline flink-runtime/src/main/java/org/apache/flink/streaming/runtime/operators/sink/committables/CheckpointCommittableManagerImpl.java:180]: This javadoc is now a bit confusing since global committer and committer follow different committing and receival strategies.

**Review by fapaul — COMMENTED** [review]
While I understand the idea of the change, I wonder if it the right time or way.

We just cut 2.0 and this is potentially a breaking change since we do not control where users append the global committer in their sink topologies. It's very theoretical and the globalcommitter was marked `@Experimental` but we should at least discuss it.

I am also a bit torn of the change of operation in the `GlobalCommitter` and the usage of the underlying infra structure i.e. `CheckpointCommittableManager`. I probably need another pass to fully understand how well the manager plays with the new semantics of the `GlobalCommitter`.

**Comment by AHeise** [comment]
> While I understand the idea of the change, I wonder if it the right time or way.
> 
> We just cut 2.0 and this is potentially a breaking change since we do not control where users append the global committer in their sink topologies. It's very theoretical and the globalcommitter was marked `@Experimental` but we should at least discuss it.
> 
> I am also a bit torn of the change of operation in the `GlobalCommitter` and the usage of the underlying infra structure i.e. `CheckpointCommittableManager`. I probably need another pass to fully understand how well the manager plays with the new semantics of the `GlobalCommitter`.

As discussed offline. The cut was just for 2.0-preview and there is going to be another cut for the actual 2.0.

File: flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/GlobalCommitterOperator.java [hunk 3]
```diff
@@ -47,10 +47,50 @@
...
+ * {@link #notifyCheckpointComplete(long)}. So by waiting for all committables of a given
+ * checkpoint, we implicitly know that the checkpoint was successful and the global committer is
+ * supposed to globally commit.
+ *
+ * <p>Note that committables of checkpoint X are not checkpointed in X because the global committer
+ * is trailing behind the checkpoint. They are replayed from the committer state in case of an
+ * error. The state only includes incomplete checkpoints coming from upstream committers not
+ * receiving {@link #notifyCheckpointComplete(long)}. All committables received are successful.
+ *
+ * <p>In rare cases, the GlobalCommitterOperator may be connected to a writer directly. In this
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/GlobalCommitterOperator.java:76]: Yes, you are right. I'll clarify the comment - the good already acknowledges that.

File: flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/GlobalCommitterOperator.java [hunk 4]
```diff
@@ -127,38 +165,67 @@ public void initializeState(StateInitializationContext context) throws Exception
...
-        if (endOfInputCommittable != null) {
-            do {
-                endOfInputCommittable.commit(committer);
-            } while (!committableCollector.isFinished());
-        }
+    private void commit() throws IOException, InterruptedException {
+        // this is true for the last commit and we need to make sure that all committables are
+        // indeed committed as this function will never be invoked again
+        boolean waitForAllCommitted =
+                lastCompletedCheckpointId == EOI
```
[the remark this answers is not in the document]
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/streaming/api/connector/sink2/GlobalCommitterOperator.java:197]: Yes that is a likely scenario. Hence the second part of this statement that will only be true iff there is an EOI committable and it has received all messages.

Then, the only way that could turn into infinite loop is when committables appear out of order (in respect to the checkpointId) which is possible with async retries. However, this applies also to the old version that loops on `endInput` (remember that we need to emit committables post EOI for the final checkpoint)

It will be solved by the next PR that removes async retries.

File: flink-runtime/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java (hunk 1 above)
Comment (AHeise) [reply flink-runtime/src/main/java/org/apache/flink/streaming/runtime/translators/GlobalCommitterTransformationTranslator.java:100]: That's a good point. What happens is: if it finds a committer, the new optimized behavior applies. If it doesn't the old unoptimized behavior is used.

So the question is what happens if you have a committer and do stuff afterwards in such a way that the committer suddenly is not doing 2PC anymore until the data reaches the global committer. I haven't been able to find such a scenario. You could have a writer->committer->union->global committer where the union is also directly fed by the writer->union->global committer but even then the output of the committer participated in 2PC.

In short, I'm convinced it's safe that any upstream committer does 2PC already and the global committer can simply piggyback.

Remember that parts of the change is that the 3PC commit protocol that we currently have is is not working properly with final checkpoints. With this fix, we get a working 2PC and "only" continue to violate endInput.

**Review by fapaul — APPROVED** [review]
Thanks for addressing the comments % please double check that ` org.apache.flink.table.planner.runtime.stream.sql.TableSinkITCase.testSinkDisorderChangeLogWithRank` doesn't fail because of this change 👍

**Comment by AHeise** [comment]
> Thanks for addressing the comments % please double check that ` org.apache.flink.table.planner.runtime.stream.sql.TableSinkITCase.testSinkDisorderChangeLogWithRank` doesn't fail because of this change 👍

I double-checked: the test uses the `SinkOperator` wrapping the old `SinkFunction` and has no changed code.

**Comment by AHeise** [comment]
@flinkbot run azure

**Comment by AHeise** [comment]
@flinkbot run azure

**Comment by AHeise** [comment]
Failure is a known issue FLINK-36591.
